core/arch/arm/tee/pta_generic.c

```diff
@@ -0,0 +1,287 @@
+	if (TEE_PARAM_TYPE_GET(param_types, idx) == TEE_PARAM_TYPE_NONE)
+		return true;
+	return false;
+}
+
+static void *alloc_transient_shm(size_t size, struct mobj **mobj)
+{
+	paddr_t p;
+	void *va;
+
+	*mobj = thread_rpc_alloc_payload(size);
+	if (!*mobj)
+		return NULL;
+
+	if (mobj_get_pa(*mobj, 0, 0, &p))
+		goto err;
+
+	if (!ALIGNMENT_IS_OK(p, uint64_t))
+		goto err;
+
+	va = mobj_get_va(*mobj, 0);
+	if (!va)
+		goto err;
+
+	return va;
+
+err:
+	thread_rpc_free_payload(*mobj);
+	return NULL;
+}
+
+static void free_transient_shm(struct mobj *mobj)
+{
+	thread_rpc_free_payload(mobj);
+}
+
+static void *prepare_memref_params(TEE_Param *param, uint32_t param_type,
+			 bool cached, struct mobj **mobj,
+			 struct thread_param *tpm)
+{
+	void *va = NULL;
+	size_t size = param->memref.size;
+
+	if (cached)
+		va = tee_fs_rpc_cache_alloc(size, mobj);
+	else
+		va = alloc_transient_shm(size, mobj);
+	if (!va)
+		return NULL;
+
+	switch (param_type) {
+	case  TEE_PARAM_TYPE_MEMREF_INPUT:
+		tpm[0] = THREAD_PARAM_MEMREF(IN, *mobj, 0, size);
+		memcpy(va, param->memref.buffer, size);
+		break;
+	case TEE_PARAM_TYPE_MEMREF_OUTPUT:
+		tpm[0] = THREAD_PARAM_MEMREF(OUT, *mobj, 0, size);
+		break;
+	case TEE_PARAM_TYPE_MEMREF_INOUT:
+		tpm[0] = THREAD_PARAM_MEMREF(INOUT, *mobj, 0, size);
+		break;
+	default:
+		goto err;
+	}
+
+	return va;
+
+err:
+	if (!cached)
+		free_transient_shm(*mobj);
+	return NULL;
+}
+
+static TEE_Result find_ree_service(void *sess_ctx, uint32_t param_types,
+				TEE_Param params[TEE_NUM_PARAMS])
+{
+	void *va;
+	struct mobj *mobj;
+	TEE_Result res = TEE_SUCCESS;
+	struct thread_param tpm[3];
+	uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+			TEE_PARAM_TYPE_VALUE_OUTPUT,
+			TEE_PARAM_TYPE_NONE,
+			TEE_PARAM_TYPE_NONE);
+
+	if (exp_pt != param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	/* Prepare RPC params for opening REE session */
+	memset(&tpm, 0, sizeof(tpm));
+
+	tpm[0] = THREAD_PARAM_VALUE(IN, OPTEE_MRC_GENERIC_OPEN, (uint64_t)sess_ctx, 0);
+
+	/* Allocate memory for passing UUID */
+	va = tee_fs_rpc_cache_alloc(params[0].memref.size, &mobj);
+	if (!va)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	tpm[1] = THREAD_PARAM_MEMREF(OUT, mobj, 0, params[0].memref.size);
+	memcpy(va, params[0].memref.buffer, params[0].memref.size);
+
+	tpm[2] = THREAD_PARAM_VALUE(OUT, 0, 0, 0);
+
+	res = thread_rpc_cmd(OPTEE_RPC_CMD_GENERIC, 3, tpm);
+	if (res == TEE_SUCCESS)
+		params[1].value.a = tpm[2].u.value.a;
+
+	return res;
+}
+
+/*
+ * Trusted Application Entry Points
+ */
+
+static TEE_Result pta_generic_open_session(uint32_t param_types __unused,
+		TEE_Param params[TEE_NUM_PARAMS] __unused,
+		void **sess_ctx)
+{
+	struct tee_ta_session *s;
+
+	/* Check that we're called from a TA */
+	s = tee_ta_get_calling_session();
+	if (!s)
+		return TEE_ERROR_ACCESS_DENIED;
+
+	*sess_ctx = (void *)(vaddr_t)get_instance_id(s);
+
+	return TEE_SUCCESS;
+}
+
+static void pta_generic_close_session(void *sess_ctx __unused)
+{
+	return;
```

Comment:
There are 4 internal commands to be implemented by REE service.
```
OPTEE_MRC_GENERIC_OPEN
OPTEE_MRC_GENERIC_CLOSE
OPTEE_MRC_GENERIC_SERVICE_START
OPTEE_MRC_GENERIC_SERVICE_STOP
```

The assumption is that REE service will implement that. So, while closing the REE service using `TEE_CloseREESession`, it will close all the handles, and there's nothing left to close at REE end.
```
void TEE_CloseREESession(TEE_REESessionHandle ree_session)
{
       TEE_Result result;
       ...
       result = TEE_InvokeTACommand(ree_session->session, 0,
                               OPTEE_MRC_GENERIC_CLOSE, paramTypes,
                               &param, NULL);
       if (result != TEE_SUCCESS)
               MSG("Failed to close the session\n");

```

Reply:
What if the TA Panics?

Reply:
There was a similar comment on the handle management of REE service, whether it should be done at PTA or API definition level. One point I wanted to add, is that it seems to be the same situation as with socket PTA. I will try to collate the discussion about TA panic in the end.